### PR TITLE
feat(op-challenger): Subscription Monitoring

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Updates the game monitoring to use a new head subscription.

**Tests**

Unit tests are performed around the tight channel loop in `MonitorGames`
using a mock subscription.

**Metadata**

Fixes CLI-4371.

### CODE REVIEW

**Comment by refcell** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #6974** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6974" target="_blank">Graphite</a> 
    * **PR #6993** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6993" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6993?utm_source=stack-comment).

File: op-challenger/fault/monitor.go [hunk 1]
```diff
@@ -143,25 +150,20 @@ func (m *gameMonitor) fetchOrCreateGamePlayer(gameData FaultDisputeGame) (gamePl
...
-	blockNum := uint64(0)
+	headers, sub, err := m.headSubscriber()
+	if err != nil {
+		return fmt.Errorf("failed to subscribe to new headers: %w", err)
+	}
+
 	for {
 		select {
+		case err := <-sub.Err():
+			return fmt.Errorf("subscription error: %w", err)
```
Comment (ajsutton) [inline op-challenger/fault/monitor.go]: I think we need to log this and reconnect the subscription. Its pretty common for a subscription to get disconnected and we don't want to crash in that situation.

File: op-challenger/fault/monitor_test.go [hunk 2]
```diff
@@ -103,10 +105,76 @@ func TestMonitorOnlyCreateSpecifiedGame(t *testing.T) {
...
+				Timestamp: 9999,
+			},
+		}
+
+		var err error
+		go func() {
+			err = monitor.MonitorGames(context.Background())
+		}()
+		mockSub.errChan <- ethereum.NotFound
+		require.ErrorIs(t, err, ethereum.NotFound)
```
Comment (ajsutton) [inline op-challenger/fault/monitor_test.go]: Also, this is a race condition depending on how quickly the go func winds up assigning to err.  I think you need a WaitGroup to allow the go func to flag when all its work is done.

File: op-challenger/fault/monitor.go (hunk 1 above)
Comment (refcell) [reply op-challenger/fault/monitor.go]: Agreed, just wanted to elicit your review for the overall setup first before going too deep in the implementation.

File: op-challenger/fault/monitor_test.go (hunk 2 above)
Comment (refcell) [reply op-challenger/fault/monitor_test.go]: Good callout hadn't paid attention to that.

File: op-challenger/fault/monitor_test.go [hunk 3]
```diff
@@ -103,10 +105,76 @@ func TestMonitorOnlyCreateSpecifiedGame(t *testing.T) {
...
+			{
+				Proxy:     addr2,
+				Timestamp: 9999,
+			},
+		}
+
+		var err error
+		go func() {
+			err = monitor.MonitorGames(context.Background())
+		}()
```
Comment (ajsutton) [inline op-challenger/fault/monitor_test.go]: Something will need to trigger this function to exit and then wait for it to complete or we'll leak resources.

File: op-service/retry/wrapper.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import (
+	"context"
+)
+
+// Wrap will continue to retry the provided function unless the
+// boolean return value is false, indicating that it should not
+// be retried. For example, if the context inside the wrapped
+// function is cancelled, it should return false so Wrap will
+// not retry it with its cancelled context.
+func Wrap(ctx context.Context, f func() (bool, error)) error {
```
Comment (ajsutton) [inline op-service/retry/wrapper.go]: This doesn't seem to use the ctx arg.

**Review by ajsutton — COMMENTED** [review]
This looks ok but the e2e tests are all failing which suggests there's a problem somewhere.  I'm a bit nervous about reinventing the persistent subscription logic that op-node already has and wonder if there's some subtle detail we've missed that the existing code would handle if we reused it.

**Comment by refcell** [comment]
Reimplementing in #7073

**Pull request closed without merging**